lang-js/src/main/java/io/crate/operation/language/JavaScriptUserDefinedFunction.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * This file is part of a module with proprietary Enterprise Features.
+ *
+ * Licensed to Crate.io Inc. ("Crate.io") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ *
+ * Unauthorized copying of this file, via any medium is strictly prohibited.
+ *
+ * To use this file, Crate.io must have given you permission to enable and
+ * use such Enterprise Features and you must have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  If you enable or use the Enterprise
+ * Features, you represent and warrant that you have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  Your use of the Enterprise Features
+ * if governed by the terms and conditions of your Enterprise or Subscription
+ * Agreement with Crate.io.
+ */
+
+package io.crate.operation.language;
+
+import io.crate.analyze.symbol.Symbol;
+import io.crate.data.Input;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.types.*;
+import jdk.nashorn.api.scripting.ScriptObjectMirror;
+import jdk.nashorn.internal.runtime.ECMAException;
+import jdk.nashorn.internal.runtime.Undefined;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.lucene.BytesRefs;
+
+import javax.script.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class JavaScriptUserDefinedFunction extends Scalar<Object, Object> {
+
+    private final FunctionInfo info;
+    private final String script;
+
+    JavaScriptUserDefinedFunction(FunctionInfo info, String script) {
+        this.info = info;
+        this.script = script;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return info;
+    }
+
+    @Override
+    public Scalar<Object, Object> compile(List<Symbol> arguments) {
+        try {
+            return new CompiledFunction(JavaScriptLanguage.bindScript(script));
+        } catch (ScriptException e) {
+            // this should not happen if the script was evaluated upfront
+            throw new UnsupportedOperationException("Could not evaluate JavaScript function", e);
+        }
+    }
+
+    @Override
+    public Object evaluate(Input<Object>[] values) {
+        try {
+            return evaluateScriptWithBindings(JavaScriptLanguage.bindScript(script), values);
+        } catch (ScriptException e) {
+            // this should not happen if the script was evaluated upfront
+            throw new UnsupportedOperationException("Could not evaluate JavaScript function", e);
```

Comment:
UnupportedOperationException is wrong, probably use some specific script evaluation exception. e.g: what do we do if builtin functions cannot be executed? 1/0?